compiler/src/iree/compiler/Codegen/Common/GPU/GPULowerToGlobalLoads.cpp

```diff
@@ -0,0 +1,338 @@
+    auto laneIdExpr = rewriter.getAffineSymbolExpr(0);
+    auto subgroupIdExpr = rewriter.getAffineSymbolExpr(1);
+    auto iterationExpr = rewriter.getAffineSymbolExpr(2);
+
+    AffineExpr strideExpr =
+        rewriter.getAffineConstantExpr(subgroupSize[0] * elementsPerCopy);
+    AffineExpr totalCopySizeExpr =
+        rewriter.getAffineConstantExpr(totalCopySizePerSubgroup);
+
+    // [subgroupStartOffset + i * gatherStride + laneId * elementsPerCopy]
+    AffineExpr gatherOffsetExpr = subgroupIdExpr * totalCopySizeExpr +
+                                  iterationExpr * strideExpr +
+                                  laneIdExpr * elementsPerCopy;
+    return affine::makeComposedAffineApply(rewriter, loc, gatherOffsetExpr,
+                                           {lIdVal, sgIdVal, indVar})
+        .getResult();
+  };
+
+  auto getSubgroupStoreBaseIndex = [&](Value sgIdVal, Value indVar) -> Value {
+    auto subgroupIdExpr = rewriter.getAffineSymbolExpr(0);
+    auto iterationExpr = rewriter.getAffineSymbolExpr(1);
+
+    AffineExpr strideExpr =
+        rewriter.getAffineConstantExpr(subgroupSize[0] * elementsPerCopy);
+    AffineExpr totalCopySizeExpr =
+        rewriter.getAffineConstantExpr(totalCopySizePerSubgroup);
+
+    // [subgroupStartOffset + i * gatherStride]
+    AffineExpr baseOffsetExpr =
+        subgroupIdExpr * totalCopySizeExpr + iterationExpr * strideExpr;
+    return affine::makeComposedAffineApply(rewriter, loc, baseOffsetExpr,
+                                           {sgIdVal, indVar})
+        .getResult();
+  };
+
+  // Build for loop skeleton:
+  scf::ForOp forOp = rewriter.create<scf::ForOp>(
+      loc, /*lb=*/rewriter.create<arith::ConstantIndexOp>(loc, 0),
+      /*ub=*/rewriter.create<arith::ConstantIndexOp>(loc, numCopiesPerThread),
+      /*steps=*/rewriter.create<arith::ConstantIndexOp>(loc, 1));
+
+  auto delinearizeIndex = [&](Value index, ArrayRef<int64_t> shape) {
+    return rewriter.create<affine::AffineDelinearizeIndexOp>(loc, index, shape)
+        .getMultiIndex();
+  };
+
+  // sync at the end of the loop across threads
+  rewriter.create<gpu::BarrierOp>(loc);
+
+  // For loop body:
+  {
+    OpBuilder::InsertionGuard guard(rewriter);
+    rewriter.setInsertionPointToStart(forOp.getBody());
+    auto inductionVar = forOp.getInductionVar();
+    auto linearizedGatherIndices =
+        getGlobalGatherIndex(subgroupId, laneId, inductionVar);
+    auto delinearizedGlobalIndices =
+        delinearizeIndex(linearizedGatherIndices, sourceType.getShape());
+    auto linearizedBaseIndices =
+        getSubgroupStoreBaseIndex(subgroupId, inductionVar);
+    auto delinearizedLocalIndices =
+        delinearizeIndex(linearizedBaseIndices, localType.getShape());
+    rewriter.create<IREE::GPU::GlobalLoadDMAOp>(
+        loc, copy.getOperand(0), delinearizedGlobalIndices,
+        copy.getOutputs()[0], delinearizedLocalIndices);
+  }
+
+  copy->erase();
```

Comment:
This sort of thing should be done via IRRewriter

And is probably a `replaceOp` - maybe a `replaceOpWithNewOp`, not an `erase`